### DISCUSSION TITLE
Add Async to ObservableDict/List/Set

### DISCUSSION THREAD

**Discussion by thetableman** [zauberzeug/nicegui#1384] [Ideas / Feature Requests]
The ObservableDict/List/Set objects are a cool tool but they don't seem to support awaitables which is a shortcoming when NiceGUI is so heavily supportive of async/await.

It seems possible that by simply adding a variation of the `handle_event` function from `events.py` might add this capability. I've done some basic testing and the below example (of the `ObservableDict` only) seems to work for both async functions and sync functions, and passing functions with parameters and without.

```py
from nicegui import background_tasks, globals
from typing import Any, Callable, Dict, Iterable, List, Set, SupportsIndex, Union, overload, Awaitable, Optional
from inspect import Parameter, signature

def handle_event(handler: Optional[Callable[..., Any]]) -> None:
    if handler is None:
        return
    try:
        expects_arguments = any(p.default is Parameter.empty and
                                p.kind is not Parameter.VAR_POSITIONAL and
                                p.kind is not Parameter.VAR_KEYWORD
                                for p in signature(handler).parameters.values())
        
        result = handler if expects_arguments else handler()
        if isinstance(result, Awaitable):
            async def wait_for_result():
                await result
            if globals.loop and globals.loop.is_running():
                background_tasks.create(wait_for_result(), name=str(handler))
            else:
                globals.app.on_startup(wait_for_result())
    except Exception as e:
        globals.handle_exception(e)

class ObservableDict(dict):

    def __init__(self, data: Dict, on_change: Callable) -> None:
        super().__init__(data)
        for key, value in self.items():
            super().__setitem__(key, make_observable(value, on_change))
        self.on_change = on_change

    def pop(self, k: Any, d: Any = None) -> Any:
        item = super().pop(k, d)
        handle_event(self.on_change)
        return item

    def popitem(self) -> Any:
        item = super().popitem()
        handle_event(self.on_change)
        return item

    def update(self, *args: Any, **kwargs: Any) -> None:
        super().update(make_observable(dict(*args, **kwargs), self.on_change))
        handle_event(self.on_change)

    def clear(self) -> None:
        super().clear()
        handle_event(self.on_change)

    def setdefault(self, __key: Any, __default: Any = None) -> Any:
        item = super().setdefault(__key, make_observable(__default, self.on_change))
        handle_event(self.on_change)
        return item

    def __setitem__(self, __key: Any, __value: Any) -> None:
        super().__setitem__(__key, make_observable(__value, self.on_change))
        handle_event(self.on_change)

    def __delitem__(self, __key: Any) -> None:
        super().__delitem__(__key)
        handle_event(self.on_change)

    def __or__(self, other: Any) -> Any:
        return super().__or__(other)

    def __ior__(self, other: Any) -> Any:
        super().__ior__(make_observable(dict(other), self.on_change))
        handle_event(self.on_change)
        return self
```

Thoughts?

**Comment by thetableman** [comment]
Also, unless I've missed it somehow, I don't think there's much (if any) documentation on the use of these objects.

**Comment by falkoschindler** [comment]
Thanks for this great suggestion, @thetableman! You're right, these observable collections are not as powerful and well documented as they could be. This is mainly because they have been developed for internal use in our storage module. But they proved to work well and could be very useful for other applications.

I think we can use the very same `handle_event` function from events.py instead of copying all this logic. It expects event arguments, but we could introduce something like an `ObservableChangeEventArguments` class for this purpose.

Since such an event isn't necessarily triggered by the user, there is no sender element and no client, making it incompatible with the `EventArguments` base class. But what about this new class hierarchy (just a rough sketch):
```py
class EventArguments:
    pass
class UiEventArguments(EventArguments):
    sender: Element
    client: Client
class ClickEventArguments(UiEventArguments):
    ...
class ObservableChangeEventArguments(EventArguments):
    ...
def handle_event(handler: Optional[Callable[..., Any]], arguments: EventArguments) -> None:
    ...
```

Then `handle_event` could check if `arguments` is an instance of `UiEventArguments` and either use the parent slot or a [`nullcontext`](https://docs.python.org/3/library/contextlib.html#contextlib.nullcontext) otherwise.

Another question is the content of `ObservableChangeEventArguments`. It might be useful to contain a reference to the observable collection and maybe the type of change that happened. But this is something we could add in a second step.

What do you think? Would you like to create a draft pull request? Then we can discuss the details there.

**Reply by thetableman** [reply]
Thanks for the positive response to the suggestion! I'll start something this week.
